File: 백준/Gold/7576. 토마토/토마토.cc
#include<iostream>
#include<queue>
#include<vector>

using namespace std;

bool range(int i,int j,int M,int N)
{
    if(i>=0 && i<N)
    {
        if(j>=0 && j<M)
            return true;
    }
    return false;
}

void bfs(vector<vector<int>> &graph,queue<pair<int,int>> tomato)
{
    int J[] = {0, 0, -1, 1};
    int I[] = {-1, 1, 0, 0};
    while (!tomato.empty())
    {
        int x = tomato.front().first;//i
        int y = tomato.front().second;//j
        tomato.pop();

        for(int i=0;i<4;++i)
        {
            int next_i = x+I[i];
            int next_j = y+J[i];
            if(range(next_i,next_j,graph[0].size(),graph.size()) && graph[next_i][next_j]==0)
            {
                graph[next_i][next_j] = graph[x][y]+1;
                tomato.push(make_pair(next_i,next_j));
            }
        }
    }
}

int main()
{
    int M,N;cin>>M>>N;
    vector<vector<int>> box(N,vector<int> (M,0));
    queue<pair<int,int>>tomato;

    for(int i=0;i<N;++i)
    {
        for(int j=0;j<M;++j)
        {
            cin>>box[i][j];
            if(box[i][j]==1)
                tomato.push(make_pair(i,j));
        }
    }
    bfs(box,tomato);

    int day = 0;

    for(int i=0;i<N;++i)
    {
        for(int j=0;j<M;++j)
        {
            if(box[i][j]==0)
            {
                cout<<-1<<endl;
                return 0;
            }
            if(day<box[i][j])
            {
                day = box[i][j];
            }
        }
    }
    cout<<day-1<<endl;
}

File: 백준/Gold/7576. 토마토/토마토.py
import sys
from collections import deque

I = [-1, 1, 0, 0]
J = [0, 0, -1, 1]

def chk(warehouse,i,j,N,M):
    if i<0 or i>=N:
        return False
    if j<0 or j>=M:
        return False
    if warehouse[i][j]==-1:
        return False
    return True

def BFS(warehouse,tomato,N,M):
    while tomato:
        now = tomato.popleft()
        for x in range (4):
            next_i = now[0]+I[x]
            next_j = now[1]+J[x]
            if chk(warehouse,next_i,next_j,N,M) and warehouse[next_i][next_j]==0:
                warehouse[next_i][next_j]= warehouse[now[0]][now[1]]+1
                tomato.append([next_i,next_j])

M,N = map(int,sys.stdin.readline().split())
warehouse = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

tomato = deque([])

for i in range (N):
    for j in range(M):
        if warehouse[i][j]==1:
            tomato.append([i,j])

BFS(warehouse, tomato, N, M)

day = 0
for i in range(N):
    for j in range(M):
        if(warehouse[i][j]==0):
            print(-1)
            exit(0)
        if(day<warehouse[i][j]):
            day = warehouse[i][j]
print(day-1)